WalletWasabi/Stores/BlockFilterSqliteStorage.cs

```diff
@@ -0,0 +1,249 @@
+	/// <exception cref="InvalidOperationException">If there is an unrecoverable error.</exception>
+	/// <seealso href="https://dev.to/lefebvre/speed-up-sqlite-with-write-ahead-logging-wal-do">Write-ahead logging explained.</seealso>
+	public static BlockFilterSqliteStorage FromFile(string dataSource, FilterModel? startingFilter = null)
+	{
+		SqliteConnectionStringBuilder builder = new();
+		builder.DataSource = dataSource;
+
+		SqliteConnection connection = new(builder.ConnectionString);
+		connection.Open();
+
+		using (SqliteCommand createCommand = connection.CreateCommand())
+		{
+			createCommand.CommandText = """
+				CREATE TABLE IF NOT EXISTS filter (
+					block_height INTEGER NOT NULL PRIMARY KEY,
+					block_hash BLOB NOT NULL,
+					filter_data BLOB NOT NULL,
+					previous_block_hash BLOB NOT NULL,
+					epoch_block_time INTEGER NOT NULL
+				)
+				""";
+			createCommand.ExecuteNonQuery();
+		}
+
+		// Enable write-ahead logging.
+		using (SqliteCommand walCommand = connection.CreateCommand())
+		{
+			walCommand.CommandText = """
+				PRAGMA journal_mode = 'wal';
+				PRAGMA synchronous  = 'NORMAL';
+				""";
+			walCommand.ExecuteNonQuery();
+		}
+
+		BlockFilterSqliteStorage storage = new(connection);
+
+		if (startingFilter is not null)
+		{
+			using SqliteCommand isEmptyCommand = connection.CreateCommand();
+			isEmptyCommand.CommandText = "SELECT count(*) FROM filter";
+			int count = Convert.ToInt32(isEmptyCommand.ExecuteScalar());
+
+			if (count == 0)
+			{
+				if (!storage.TryAppend(startingFilter))
+				{
+					// Unrecoverable error.
+					throw new InvalidOperationException("Failed to add the first filter to the database.");
+				}
+			}
+		}
+
+		return storage;
+	}
+
+	/// <summary>
+	/// Begin a transaction.
+	/// </summary>
+	/// <seealso href="https://learn.microsoft.com/en-us/dotnet/standard/data/sqlite/transactions"/>
+	public SqliteTransaction BeginTransaction()
+	{
+		return Connection.BeginTransaction();
+	}
+
+	/// <summary>
+	/// Returns all filters with height ≥ than the given one.
+	/// </summary>
+	public IEnumerable<FilterModel> Fetch(int fromHeight)
+	{
+		using SqliteCommand command = Connection.CreateCommand();
+		command.CommandText = "SELECT * FROM filter WHERE block_height >= $block_height ORDER BY block_height";
+		command.Parameters.AddWithValue("$block_height", fromHeight);
+
+		using SqliteDataReader reader = command.ExecuteReader();
+
+		while (reader.Read())
+		{
+			FilterModel filter = ReadRow(reader);
+			yield return filter;
+		}
+	}
+
+	/// <summary>
+	/// Returns last <paramref name="n"/> filters from the database table.
+	/// </summary>
+	public IEnumerable<FilterModel> FetchLast(int n)
+	{
+		using SqliteCommand command = Connection.CreateCommand();
+		command.CommandText = "SELECT * FROM filter WHERE block_height >= (SELECT MAX(block_height) - $n FROM filter) ORDER BY block_height";
```

Comment:
This should be faster, no subselect. Not sure that parameter thing will work for `LIMIT` clause, maybe it should be appened to command itself.
```suggestion
		command.CommandText = "SELECT * FROM filter ORDER BY block_height DESC LIMIT $n";
```

Reply:
I guess `dotnet trace` should be useful here, to see what code takes most of the time.